<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>breeDBase • breedbase</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="breeDBase">
<meta property="og:description" content="Helper functions for generating breeDBase upload templates.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">breedbase</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="breedbaser">breedbase.R<a class="anchor" aria-label="anchor" href="#breedbaser"></a>
</h1></div>
<div class="section level3">
<h3 id="an-r-package-for-generating-breedbase-upload-templates">An R package for generating breedbase upload templates<a class="anchor" aria-label="anchor" href="#an-r-package-for-generating-breedbase-upload-templates"></a>
</h3>
<p>This R package can be used to create classes of various breedbase data types (such as an Accession, a phenotyping Trial, a Plot of a phenotyping Trial, etc). One or more instances of a class can be passed to a <code>buildTemplate</code> or <code>writeTemplate</code> function to create and/or write an upload template to be used for adding data through the breedbase website.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Currently, the breedbase package can be installed directly from GitHub using <code>devtools</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install the devtools package, if you don't have it already</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install the breedbase package directly from GitHub</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"TriticeaeToolbox/breedbase.R"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">breedbase</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>Documentation for all Classes and functions can be found in the package’s R documentation or online at <a href="https://triticeaetoolbox.github.io/breedbase.R/reference" class="external-link">https://triticeaetoolbox.github.io/breedbase.R/reference</a>.</p>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<div class="section level3">
<h3 id="accessions-and-pedigrees">Accessions and Pedigrees<a class="anchor" aria-label="anchor" href="#accessions-and-pedigrees"></a>
</h3>
<div class="section level4">
<h4 id="accessions">Accessions<a class="anchor" aria-label="anchor" href="#accessions"></a>
</h4>
<p>The <code>Accession</code> Class holds all of the information about a single Accession. The <code><a href="reference/Accession.html">Accession()</a></code> function can be used to create an instance of the <code>Accession</code> Class. The <code>accession_name</code> and <code>species_name</code> and required parameters. Additional optional parameters can be provided through a named list passed as the <code>properties</code> parameter. For a list of all of the supported optional parameters, see the <a href="https://triticeaetoolbox.github.io/breedbase.R/reference/Accession-class.html" class="external-link">Accession Class Documentation</a> or use the <code><a href="reference/getSupportedAccessionProperties.html">getSupportedAccessionProperties()</a></code> function to get a list of the names of all of the supported properties.</p>
<p><strong>NOTE:</strong> With version 1.0.0 of the package, the Accession functions now use the new updated breedbase names for the Accession properties. The old names are no longer supported by the package.</p>
<p>In order to avoid duplicate Accession entries in the database, it is recommended to perform a search of your Accession names against the names of those already in the database to find different possible variations in the spelling of the Accession name. The package now includes functions that interface with the <a href="https://github.com/TriticeaeToolbox/BrAPI-germplasm-search" class="external-link">BrAPI germplasm search tool</a> so Accession name searches can now be done directly in R. See the <a href="#accession-search">Accession Search</a> section below for more information on how to perform the search using this package.</p>
<p><strong>Example:</strong> Create new Accessions with some optional properties</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">jerry</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Accession.html">Accession</a></span><span class="op">(</span></span>
<span>     accession_name <span class="op">=</span> <span class="st">"JERRY"</span>, </span>
<span>     species_name <span class="op">=</span> <span class="st">"Triticum aestivum"</span>,</span>
<span>     properties <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>         synonym <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ND9257"</span>, <span class="st">"PI632433"</span><span class="op">)</span>,</span>
<span>         <span class="st">"institute code"</span> <span class="op">=</span> <span class="st">"NDSU"</span>,</span>
<span>         organization_name <span class="op">=</span> <span class="st">"North Dakota State University"</span></span>
<span>     <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">caledonia</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Accession.html">Accession</a></span><span class="op">(</span></span>
<span>     accession_name <span class="op">=</span> <span class="st">"CALEDONIA"</span>, </span>
<span>     species_name <span class="op">=</span> <span class="st">"Triticum aestivum"</span>,</span>
<span>     properties <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>         synonym <span class="op">=</span> <span class="st">"PI610188"</span>,</span>
<span>         <span class="st">"institute code"</span> <span class="op">=</span> <span class="st">"CNL"</span>,</span>
<span>         organization_name <span class="op">=</span> <span class="st">"Cornell University"</span></span>
<span>     <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_cross</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Accession.html">Accession</a></span><span class="op">(</span></span>
<span>    accession_name <span class="op">=</span> <span class="st">"MY_CROSS"</span>,</span>
<span>    species_name <span class="op">=</span> <span class="st">"Triticum aestivum"</span>,</span>
<span>    properties <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"institute code"</span> <span class="op">=</span> <span class="st">"CNL"</span>,</span>
<span>        organization_name <span class="op">=</span> <span class="st">"Cornell University"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add all accessions to a vector</span></span>
<span><span class="va">accessions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">jerry</span>, <span class="va">caledonia</span>, <span class="va">my_cross</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pedigrees">Pedigrees<a class="anchor" aria-label="anchor" href="#pedigrees"></a>
</h4>
<p>The <code><a href="reference/Pedigree.html">Pedigree()</a></code> function can be used to create an object of the <code>Pedigree</code> Class which sets the parents (and optionally the cross type) of a single Accession.</p>
<p><strong>Example:</strong> Create a <code>Pedigree</code> object that sets the parents of the <code>my_cross</code> Accession as <code>jerry</code> and <code>caledonia</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_cross_pedigree</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Pedigree.html">Pedigree</a></span><span class="op">(</span><span class="va">my_cross</span>, female_parent <span class="op">=</span> <span class="va">jerry</span>, male_parent <span class="op">=</span> <span class="va">caledonia</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> The <strong>T3/Wheat</strong> breedbase instance is planning on keeping pedigree information stored as an unparsed purdy pedigree string (at least until we have time to try to parse purdy pedigrees into parents). The purdy pedigree string will be stored as an optional property (<code>purdy_pedigree</code>) of the Accession and can be added to the Accession upload template.</p>
</div>
<div class="section level4">
<h4 id="creating-upload-templates">Creating Upload Templates<a class="anchor" aria-label="anchor" href="#creating-upload-templates"></a>
</h4>
<p>Once you have one or more objects of a particular Class (such as a vector of <code>Accession</code>s or a <code>Pedigree</code>) you can create the create the upload templates used to add that particular data type to a breedbase database through its website.</p>
<p><strong>Example:</strong> Create an upload template for the 3 Accessions and 1 Pedigree.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/writeAccessionTemplate.html">writeAccessionTemplate</a></span><span class="op">(</span><span class="va">accessions</span>, <span class="st">'/path/to/accessions.xls'</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/writePedigreeTemplate.html">writePedigreeTemplate</a></span><span class="op">(</span><span class="va">my_cross_pedigree</span>, <span class="st">"/path/to/pedigrees.txt"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="locations">Locations<a class="anchor" aria-label="anchor" href="#locations"></a>
</h3>
<p>Locations are required to have latitude and longitude coordinates as well as their altitude/elevation. These can be added manually as parameters or geocoded using the <code><a href="reference/geocodeLocation.html">geocodeLocation()</a></code> function which can lookup a street address or town name. The function uses the <a href="https://nominatim.org/release-docs/develop/api/Overview/" class="external-link">OpenStreetMap Nominatim API</a> for geocoding and the <a href="https://www.gmrt.org/services/pointserverinfo.php#!/services/getGMRTPoint" class="external-link">Global Multi-Resolution Topography PointServer Web Service</a> for elevation data.</p>
<p>Additionally, Locations are required to have a NOAA Station ID, which is used to get historical weather records. The <code>lookupNOAAStationId()</code> function can be used to lookup a weather station close to the Location.</p>
<p>Use the <code><a href="reference/getCountryCodes.html">getCountryCodes()</a></code> function to get a list of the supported country codes and the <code><a href="reference/getLocationTypes.html">getLocationTypes()</a></code> function to get a list of the supported location types.</p>
<p><strong>Example:</strong> Create a Location with all of the required parameters known</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loc1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Location.html">Location</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Batavia, NY"</span>,</span>
<span>    abbreviation <span class="op">=</span> <span class="st">"BAT"</span>,</span>
<span>    country_code <span class="op">=</span> <span class="st">"USA"</span>,</span>
<span>    country_name <span class="op">=</span> <span class="st">"United States of America"</span>,</span>
<span>    program <span class="op">=</span> <span class="st">"Cornell"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"Field"</span>,</span>
<span>    latitude <span class="op">=</span> <span class="fl">42.98054</span>,</span>
<span>    longitude <span class="op">=</span> <span class="op">-</span><span class="fl">78.23176</span>,</span>
<span>    altitude <span class="op">=</span> <span class="fl">274</span>,</span>
<span>    noaa_station_id <span class="op">=</span> <span class="st">"GHCND:US1NYGN0013"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Example:</strong> Create a Location without knowing its lat, lon or elevation/altitude.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loc2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Location.html">Location</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Batavia, NY"</span>,</span>
<span>    abbreviation <span class="op">=</span> <span class="st">"BAT"</span>,</span>
<span>    country_code <span class="op">=</span> <span class="st">"USA"</span>,</span>
<span>    country_name <span class="op">=</span> <span class="st">"United States of America"</span>,</span>
<span>    program <span class="op">=</span> <span class="st">"Cornell"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"Field"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>If either the <code>latitude</code>, <code>longitude</code> or <code>altitude</code> parameters are not provided, they will be queried using the <code>name</code> as the location.</li>
<li>If the <code>noaa_station_id</code> is not provided, it will be queried using the provided or queried lat and lon coordinates of the location.</li>
</ul>
<p><strong>Example:</strong> Geocode a specific address first, then use that information to create the Location</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/geocodeLocation.html">geocodeLocation</a></span><span class="op">(</span><span class="st">"2 Caldwell Drive, Ithaca, NY"</span><span class="op">)</span></span>
<span><span class="va">loc3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Location.html">Location</a></span><span class="op">(</span></span>
<span>     name <span class="op">=</span> <span class="st">"Ithaca, NY - Caldwell"</span>, </span>
<span>     abbreviation <span class="op">=</span> <span class="st">"ITHNY-CAL"</span>,</span>
<span>     country_code <span class="op">=</span> <span class="st">"USA"</span>,</span>
<span>     country_name <span class="op">=</span> <span class="st">"United States of America"</span>,</span>
<span>     program <span class="op">=</span> <span class="st">"Cornell"</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"Field"</span>,</span>
<span>     latitude <span class="op">=</span> <span class="va">geo</span><span class="op">$</span><span class="va">latitude</span>,</span>
<span>     longitude <span class="op">=</span> <span class="va">geo</span><span class="op">$</span><span class="va">longitude</span>,</span>
<span>     altitude <span class="op">=</span> <span class="va">geo</span><span class="op">$</span><span class="va">altitude</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can use the <code>lookupNOAAStationId()</code> function to get ID of the closest NOAA weather station to the specified geographic coordinates.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stationId</span> <span class="op">&lt;-</span> <span class="fu">lookupNOAAStationId</span><span class="op">(</span><span class="va">geo</span><span class="op">$</span><span class="va">latitude</span>, <span class="va">geo</span><span class="op">$</span><span class="va">longitude</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="trials">Trials<a class="anchor" aria-label="anchor" href="#trials"></a>
</h3>
<p>A <code>Trial</code> includes the metadata about a single phenotyping trial as well as information about the individual plots in the trial (see the <a href="#plots">Plots</a> section below). A trial is assigned a unique name, associated with a breeding program, year and location, and includes the trial design type and a description. Additional information can include the field and plot sizes, planting and harvest dates, and type of trial (greenhouse, phenotyping, advanced yield, etc).</p>
<p>A <code>Trial</code> object can be created without its plot information (it just contains metadata about the trial). However, the plot information must be added to the <code>Trial</code> before a trial upload template can be created.</p>
<p><strong>Example:</strong> Create a Trial with just the required parameters</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trial_mad</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Trial.html">Trial</a></span><span class="op">(</span></span>
<span>    trial_name <span class="op">=</span> <span class="st">"UMOPN_2019_Madison"</span>, </span>
<span>    breeding_program <span class="op">=</span> <span class="st">"University of Wisconsin"</span>, </span>
<span>    location <span class="op">=</span> <span class="st">"Madison, WI"</span>, </span>
<span>    year <span class="op">=</span> <span class="fl">2019</span>, </span>
<span>    design_type <span class="op">=</span> <span class="st">"RCBD"</span>, </span>
<span>    description <span class="op">=</span> <span class="st">"UMOPN Nursery Trial"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Example:</strong> Create a Trial with optional parameters</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">opts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>planting_date <span class="op">=</span> <span class="st">"2019-04-25"</span>, harvest_date <span class="op">=</span> <span class="st">"2019-10-05"</span>, trial_type <span class="op">=</span> <span class="st">"phenotyping_trial"</span><span class="op">)</span></span>
<span><span class="va">trial_arl</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Trial.html">Trial</a></span><span class="op">(</span></span>
<span>    trial_name <span class="op">=</span> <span class="st">"UMOPN_2019_Arlington"</span>, </span>
<span>    breeding_program <span class="op">=</span> <span class="st">"University of Wisconsin"</span>, </span>
<span>    location <span class="op">=</span> <span class="st">"Arlington, WI"</span>, </span>
<span>    year <span class="op">=</span> <span class="fl">2019</span>, </span>
<span>    design_type <span class="op">=</span> <span class="st">"RCBD"</span>, </span>
<span>    description <span class="op">=</span> <span class="st">"UMOPN Nursery Trial"</span>, </span>
<span>    properties <span class="op">=</span> <span class="va">opts</span></span>
<span><span class="op">)</span> </span></code></pre></div>
<p>See the <a href="https://triticeaetoolbox.github.io/breedbase.R/reference/Trial-class.html" class="external-link">Trial Class Documentation</a> for more information on the optional parameters.</p>
<p>Use the <code><a href="reference/getTrialDesignTypes.html">getTrialDesignTypes()</a></code> function to get list of supported Trial design types and the <code><a href="reference/getTrialTypes.html">getTrialTypes()</a></code> function to get a list of supported Trial types.</p>
<p><strong>Adding Plots:</strong> Once the <code>Plot</code>s for the trial have been created, they can be added to the <code>Trial</code> with the <code><a href="reference/setTrialPlots.html">setTrialPlots()</a></code> function. See the <a href="#plots">Plots section below</a> for more information on creating Trial Plots.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trial_mad</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/setTrialPlots.html">setTrialPlots</a></span><span class="op">(</span><span class="va">trial_mad</span>, <span class="va">plots_mad</span><span class="op">)</span></span>
<span><span class="va">trial_arl</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/setTrialPlots.html">setTrialPlots</a></span><span class="op">(</span><span class="va">trial_arl</span>, <span class="va">plots_arl</span><span class="op">)</span></span></code></pre></div>
<p><strong>Creating Trial Templates:</strong> Once you have one or more <code>Trial</code>s (that have <code>Plots</code> set), you can create a trial upload template that can be submitted to a breedbase website from the <strong>Manage Trials</strong> page (‘Upload Existing Trials’ &gt; ‘Multiple Trial Designs’).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/writeTrialTemplate.html">writeTrialTemplate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">trial_mad</span>, <span class="va">trial_arl</span><span class="op">)</span>, <span class="st">'/path/to/trials.xls'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h3>
<p>A series of Plots are used to describe a phenotyping trial layout. Each plot is assigned a unique name, a plot number, the name of the Accession used in the plot, and the block it is located in. If the physical layout of the plots is known, each Plot can be assigned a row and column number.</p>
<ul><li>A plot layout can be generated by manually creating each Plot with the desired parameters</li></ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create the Plots</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Plot.html">Plot</a></span><span class="op">(</span></span>
<span>    plot_name <span class="op">=</span> <span class="st">"FARM-2019-UNH_PLOT1"</span>, </span>
<span>    accession_name <span class="op">=</span> <span class="st">"SL18-UCONN-S131"</span>, </span>
<span>    plot_number <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    block_number <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    properties <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        row_numer <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        col_number <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        rep_number <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Plot.html">Plot</a></span><span class="op">(</span></span>
<span>    plot_name <span class="op">=</span> <span class="st">"FARM-2019-UNH_PLOT2"</span>, </span>
<span>    accession_name <span class="op">=</span> <span class="st">"SL18-UCONN-S31"</span>, </span>
<span>    plot_number <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    block_number <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    properties <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        row_numer <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        col_number <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        rep_number <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine the Plots for the layout</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Write the Plot layout template</span></span>
<span><span class="fu">writePlotLayout</span><span class="op">(</span><span class="va">plots</span>, <span class="st">"/path/to/plots.xls"</span><span class="op">)</span></span></code></pre></div>
<ul><li>Alternatively, a simple plot layout can be generated automatically when given a list of Accessions and basic layout properties.</li></ul>
<p><strong>Example</strong>: A trial with 24 plots that has a maximum of 6 columns. The blocks have a dimension of 3 columns by 2 rows. The Plots will start in the top left corner and work across columns and down rows.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the list of Accessions</span></span>
<span><span class="co"># here we're using a dummy list of Accessions named ACC_A - ACC_X</span></span>
<span><span class="va">accessions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">LETTERS</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">24</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="reference/Accession.html">Accession</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ACC_"</span>, <span class="va">x</span><span class="op">)</span>, <span class="st">"Saccharina latissima"</span><span class="op">)</span> <span class="op">}</span><span class="op">)</span></span>
<span>   </span>
<span><span class="co"># Create the Plots</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/createPlots.html">createPlots</a></span><span class="op">(</span></span>
<span>    trial_name <span class="op">=</span> <span class="st">"TEST-TRIAL"</span>, </span>
<span>    accessions <span class="op">=</span> <span class="va">accessions</span>, </span>
<span>    max_cols <span class="op">=</span> <span class="fl">6</span>, </span>
<span>    max_cols_per_block <span class="op">=</span> <span class="fl">3</span>, </span>
<span>    max_rows_per_block <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># The generated layout:</span></span>
<span><span class="co">#                                     Col1                Col2                Col3                Col4                Col5                Col6</span></span>
<span><span class="co"># ===== Row1 =====      ===== Plot 1 =====  ===== Plot 2 =====  ===== Plot 3 =====  ===== Plot 4 =====  ===== Plot 5 =====  ===== Plot 6 =====</span></span>
<span><span class="co"># Row1: Plot Name         TEST-TRIAL_PLOT1    TEST-TRIAL_PLOT2    TEST-TRIAL_PLOT3    TEST-TRIAL_PLOT4    TEST-TRIAL_PLOT5    TEST-TRIAL_PLOT6</span></span>
<span><span class="co"># Row1: Accession Name               ACC_A               ACC_B               ACC_C               ACC_D               ACC_E               ACC_F</span></span>
<span><span class="co"># Row1: Block                            1                   1                   1                   2                   2                   2</span></span>
<span><span class="co"># Row1: Control                      FALSE               FALSE               FALSE               FALSE               FALSE               FALSE</span></span>
<span><span class="co"># ===== Row2 =====      ===== Plot 7 =====  ===== Plot 8 =====  ===== Plot 9 ===== ===== Plot 10 ===== ===== Plot 11 ===== ===== Plot 12 =====</span></span>
<span><span class="co"># Row2: Plot Name         TEST-TRIAL_PLOT7    TEST-TRIAL_PLOT8    TEST-TRIAL_PLOT9   TEST-TRIAL_PLOT10   TEST-TRIAL_PLOT11   TEST-TRIAL_PLOT12</span></span>
<span><span class="co"># Row2: Accession Name               ACC_G               ACC_H               ACC_I               ACC_J               ACC_K               ACC_L</span></span>
<span><span class="co"># Row2: Block                            1                   1                   1                   2                   2                   2</span></span>
<span><span class="co"># Row2: Control                      FALSE               FALSE               FALSE               FALSE               FALSE               FALSE</span></span>
<span><span class="co"># ===== Row3 =====     ===== Plot 13 ===== ===== Plot 14 ===== ===== Plot 15 ===== ===== Plot 16 ===== ===== Plot 17 ===== ===== Plot 18 =====</span></span>
<span><span class="co"># Row3: Plot Name        TEST-TRIAL_PLOT13   TEST-TRIAL_PLOT14   TEST-TRIAL_PLOT15   TEST-TRIAL_PLOT16   TEST-TRIAL_PLOT17   TEST-TRIAL_PLOT18</span></span>
<span><span class="co"># Row3: Accession Name               ACC_M               ACC_N               ACC_O               ACC_P               ACC_Q               ACC_R</span></span>
<span><span class="co"># Row3: Block                            3                   3                   3                   4                   4                   4</span></span>
<span><span class="co"># Row3: Control                      FALSE               FALSE               FALSE               FALSE               FALSE               FALSE</span></span>
<span><span class="co"># ===== Row4 =====     ===== Plot 19 ===== ===== Plot 20 ===== ===== Plot 21 ===== ===== Plot 22 ===== ===== Plot 23 ===== ===== Plot 24 =====</span></span>
<span><span class="co"># Row4: Plot Name        TEST-TRIAL_PLOT19   TEST-TRIAL_PLOT20   TEST-TRIAL_PLOT21   TEST-TRIAL_PLOT22   TEST-TRIAL_PLOT23   TEST-TRIAL_PLOT24</span></span>
<span><span class="co"># Row4: Accession Name               ACC_S               ACC_T               ACC_U               ACC_V               ACC_W               ACC_X</span></span>
<span><span class="co"># Row4: Block                            3                   3                   3                   4                   4                   4</span></span>
<span><span class="co"># Row4: Control                      FALSE               FALSE               FALSE               FALSE               FALSE               FALSE </span></span></code></pre></div>
<p><strong>Example:</strong> The same 24 plots, but the plots are assigned using a zig-zag method (ie left to right in the first row, right to left in the second row, etc). Additionally, the controls are specified using the names of the Accessions used as controls.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the list of Accessions</span></span>
<span><span class="va">accessions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">LETTERS</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">24</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="reference/Accession.html">Accession</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ACC_"</span>, <span class="va">x</span><span class="op">)</span>, <span class="st">"Saccharina latissima"</span><span class="op">)</span> <span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the Plots</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/createPlots.html">createPlots</a></span><span class="op">(</span></span>
<span>    trial_name <span class="op">=</span> <span class="st">"TEST-TRIAL"</span>, </span>
<span>    accessions <span class="op">=</span> <span class="va">accessions</span>, </span>
<span>    max_cols <span class="op">=</span> <span class="fl">6</span>, </span>
<span>    max_cols_per_block <span class="op">=</span> <span class="fl">3</span>, </span>
<span>    max_rows_per_block <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    zig_zag <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    controls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ACC_D"</span>, <span class="st">"ACC_L"</span>, <span class="st">"ACC_O"</span>, <span class="st">"ACC_T"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The generated layout:</span></span>
<span><span class="co"># ===== Row1 =====      ===== Plot 1 =====  ===== Plot 2 =====  ===== Plot 3 =====  ===== Plot 4 =====  ===== Plot 5 =====  ===== Plot 6 =====</span></span>
<span><span class="co"># Row1: Plot Name         TEST-TRIAL_PLOT1    TEST-TRIAL_PLOT2    TEST-TRIAL_PLOT3    TEST-TRIAL_PLOT4    TEST-TRIAL_PLOT5    TEST-TRIAL_PLOT6</span></span>
<span><span class="co"># Row1: Accession Name               ACC_A               ACC_B               ACC_C               ACC_D               ACC_E               ACC_F</span></span>
<span><span class="co"># Row1: Block                            1                   1                   1                   2                   2                   2</span></span>
<span><span class="co"># Row1: Control                      FALSE               FALSE               FALSE                TRUE               FALSE               FALSE</span></span>
<span><span class="co"># ===== Row2 =====     ===== Plot 12 ===== ===== Plot 11 ===== ===== Plot 10 =====  ===== Plot 9 =====  ===== Plot 8 =====  ===== Plot 7 =====</span></span>
<span><span class="co"># Row2: Plot Name        TEST-TRIAL_PLOT12   TEST-TRIAL_PLOT11   TEST-TRIAL_PLOT10    TEST-TRIAL_PLOT9    TEST-TRIAL_PLOT8    TEST-TRIAL_PLOT7</span></span>
<span><span class="co"># Row2: Accession Name               ACC_L               ACC_K               ACC_J               ACC_I               ACC_H               ACC_G</span></span>
<span><span class="co"># Row2: Block                            1                   1                   1                   2                   2                   2</span></span>
<span><span class="co"># Row2: Control                       TRUE               FALSE               FALSE               FALSE               FALSE               FALSE</span></span>
<span><span class="co"># ===== Row3 =====     ===== Plot 13 ===== ===== Plot 14 ===== ===== Plot 15 ===== ===== Plot 16 ===== ===== Plot 17 ===== ===== Plot 18 =====</span></span>
<span><span class="co"># Row3: Plot Name        TEST-TRIAL_PLOT13   TEST-TRIAL_PLOT14   TEST-TRIAL_PLOT15   TEST-TRIAL_PLOT16   TEST-TRIAL_PLOT17   TEST-TRIAL_PLOT18</span></span>
<span><span class="co"># Row3: Accession Name               ACC_M               ACC_N               ACC_O               ACC_P               ACC_Q               ACC_R</span></span>
<span><span class="co"># Row3: Block                            3                   3                   3                   4                   4                   4</span></span>
<span><span class="co"># Row3: Control                      FALSE               FALSE                TRUE               FALSE               FALSE               FALSE</span></span>
<span><span class="co"># ===== Row4 =====     ===== Plot 24 ===== ===== Plot 23 ===== ===== Plot 22 ===== ===== Plot 21 ===== ===== Plot 20 ===== ===== Plot 19 =====</span></span>
<span><span class="co"># Row4: Plot Name        TEST-TRIAL_PLOT24   TEST-TRIAL_PLOT23   TEST-TRIAL_PLOT22   TEST-TRIAL_PLOT21   TEST-TRIAL_PLOT20   TEST-TRIAL_PLOT19</span></span>
<span><span class="co"># Row4: Accession Name               ACC_X               ACC_W               ACC_V               ACC_U               ACC_T               ACC_S</span></span>
<span><span class="co"># Row4: Block                            3                   3                   3                   4                   4                   4</span></span>
<span><span class="co"># Row4: Control                      FALSE               FALSE               FALSE               FALSE                TRUE               FALSE</span></span></code></pre></div>
<p><strong>Creating Plot Templates:</strong> Once you have a vector of <code>Plot</code>s for a single trial, a plot upload template can be created. This upload template can be used to create a single Trial from the <strong>Manage Trials</strong> page (‘Upload Existing Trial(s)’ &gt; ‘Single Trial Design’). Here you will enter the Trial metadata in a form and upload this template for the trial’s plot layout.</p>
</div>
<div class="section level3">
<h3 id="plot-data">Plot Data<a class="anchor" aria-label="anchor" href="#plot-data"></a>
</h3>
<p>The <code>PlotData</code> class is used to represent the phenotype observations (one or more traits) of a single Plot from a single Trial. A vector of <code>PlotData</code> objects can be used to represent the observations from an entire Trial and can be used to create a plot data upload template that can be uploaded through a breedbase website.</p>
<p>A <code>PlotData</code> object is created with a <code>plot_name</code> and a named list of <code>observations</code> (and optionally <code>notes</code> for the plot).</p>
<p>The <code>plot_name</code> must match the name of an existing Plot in an existing Trial.</p>
<p>The <code>observations</code> must take the form of a named list where the list item name/key is the full trait name of the trait observed and the list item value is the observed trait value. The trait name must match an existing trait name in the breedbase database and follow the form {trait name}|{trait id} (ex: <code>Grain yield kg/ha|CO_321:0001218</code>).</p>
<p><strong>Example:</strong> Create two <code>PlotData</code> objects representing two plots from the same Trial</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plotData1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PlotData.html">PlotData</a></span><span class="op">(</span></span>
<span>    <span class="st">"FARM-2019-UNH_PLOT1"</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"Blade length cm|CO_360:0000240"</span> <span class="op">=</span> <span class="fl">24</span>, </span>
<span>        <span class="st">"Blade width cm|CO_360:0000241"</span> <span class="op">=</span> <span class="fl">5</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plotData2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PlotData.html">PlotData</a></span><span class="op">(</span></span>
<span>    <span class="st">"FARM-2019-UNH_PLOT2"</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"Blade length cm|CO_360:0000240"</span> <span class="op">=</span> <span class="fl">27</span>, </span>
<span>        <span class="st">"Blade width cm|CO_360:0000241"</span> <span class="op">=</span> <span class="fl">4</span>, </span>
<span>        <span class="st">"Blade thickness mm|CO_360:0000248"</span> <span class="op">=</span> <span class="fl">2</span></span>
<span>    <span class="op">)</span>, </span>
<span>    notes<span class="op">=</span><span class="st">"This plot showed signs of disturbance"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">trialPlotData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">plotData1</span>, <span class="va">plotData2</span><span class="op">)</span></span></code></pre></div>
<p><strong>Creating Plot Data Templates:</strong> Once you have a vector of <code>PlotData</code> objects for a single trial, a plot data upload template can be created using the <code><a href="reference/writePlotDataTemplate.html">writePlotDataTemplate()</a></code> function. This will create a a phenotyping results spreadsheet in the ‘Simple’ format with plot-level data. This file can be uploaded from the detail page of a specific Trial (under ‘Upload Data Files’ &gt; ‘Phenotyping Spreadsheets’) or from the ‘Manage Phenotypic Data’ page.</p>
</div>
</div>
<div class="section level2">
<h2 id="conversion-functions">Conversion Functions<a class="anchor" aria-label="anchor" href="#conversion-functions"></a>
</h2>
<p>A number of helpful conversion functions are included to help convert phenotype observation values between different commonly used units. All of these functions start with <code>convert_</code>, such as <code>convert_buac_kgHa</code>. See the package Documentation for more details.</p>
</div>
<div class="section level2">
<h2 id="accession-search">Accession Search<a class="anchor" aria-label="anchor" href="#accession-search"></a>
</h2>
<p>The package now includes functions that can interface with a <a href="https://synonyms.triticeaetoolbox.org" class="external-link">BrAPI germplasm search</a> server. The search tool allows the user to find matching database entries that are the same as or similar to the user-provided search terms using a number of configurable search options. Performing a search before submitting new Accessions to a breedbase database can limit the number of accidental duplicate entries in the database.</p>
<p>To perform a search, you first need to create an Accession Search Database object which contains the information about the breedbase instance you are going to search on. The Accession search server has a number of pre-configured databases you can use (where the connection information is set by the search server) or you can create your own database specification.</p>
<p>To get and choose a pre-configured database from the search server:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a><span class="co"># Get a list of supported databases</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a>dbs &lt;-<span class="st"> </span><span class="kw">getAccessionSearchDBs</span>()</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">"1 = T3/Wheat: https://wheat.triticeaetoolbox.org/brapi/v1 [v1.3]"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">"2 = T3/Oat: https://oat.triticeaetoolbox.org/brapi/v1 [v1.3]"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">"3 = T3/Barley: https://barley.triticeaetoolbox.org/brapi/v1 [v1.3]"</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true"></a>[<span class="dv">1</span>] <span class="st">"4 = Cassavabase: https://cassavabase.org/brapi/v1 [v1.3]"</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true"></a><span class="co"># Select a database to use</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true"></a>db &lt;-<span class="st"> </span>dbs[[<span class="dv">1</span>]]</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true"></a><span class="co"># Or get the Database by name</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true"></a>db &lt;-<span class="st"> </span><span class="kw">getAccessionSearchDB</span>(<span class="st">"T3/Wheat"</span>)</span></code></pre></div>
<p>To create your own database connection:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/createAccessionSearchDB.html">createAccessionSearchDB</a></span><span class="op">(</span><span class="st">"T3/Wheat-Sandbox"</span>, <span class="st">"https://wheat-sandbox.triticeaetoolbox.org/brapi/v1"</span>, <span class="st">"v1.3"</span><span class="op">)</span></span></code></pre></div>
<p>Once you have a configured database connection, you can update the cache of database terms that the search server uses:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/updateAccessionSearchCache.html">updateAccessionSearchCache</a></span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span></code></pre></div>
<p>Then, you can perform a search on a vector of your search terms:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">search_terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jerry"</span>, <span class="st">"SY-Gold"</span>, <span class="st">"PU0128A1_36"</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/performAccessionSearch.html">performAccessionSearch</a></span><span class="op">(</span><span class="va">db</span>, <span class="va">search_terms</span><span class="op">)</span> </span></code></pre></div>
<p>The results are returned as a tibble with a column containing your search term and information about the matches:</p>
<pre><code># A tibble: 3 x 6
  search_term search_routine     germplasm_name germplasm_id database_term database_term_type
  &lt;chr&gt;       &lt;chr&gt;              &lt;chr&gt;                 &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;             
1 jerry       Exact Match        JERRY                230227 JERRY         name              
2 SY-Gold     Exact Match        00X0100-51           232633 Sy-Gold       synonym           
3 PU0128A1_36 Remove Punctuation 0128A1-36            238801 PU0128A1-36   synonym     </code></pre>
</div>
<div class="section level2">
<h2 id="configurable-options">Configurable Options<a class="anchor" aria-label="anchor" href="#configurable-options"></a>
</h2>
<p>The breedbase package has a number of default property values that can be overriden using R’s global options. The following table lists the options that can be changed. To set the value of an option, use the <code><a href="https://rdrr.io/r/base/options.html" class="external-link">options()</a></code> function with the name of the option and the new value (Example: to change the supported Accession properties: <code>options("breedbase.editable_stock_props" = c("variety", "class", "notes", "purdy_pedigree"))</code></p>
<table class="table">
<colgroup>
<col width="14%">
<col width="37%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="left">Name</th>
<th align="left">Definition                                              </th>
<th align="left">Default Value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">breedbase.standard_stock_props</td>
<td align="left">Accession properties that are used by the breedbase website</td>
<td align="left"><code>c("population_name", "organization_name", "synonym", "PUI")</code></td>
</tr>
<tr class="even">
<td align="left">breedbase.editable_stock_props</td>
<td align="left">Additional Accession properties defined by the individual breedbase instance</td>
<td align="left"><code>c("variety", "released_variety_name", "donor", "donor institute", "donor PUI", "country of origin", "state", "institute code", "institute name", "biological status of accession code", "notes", "accession number", "seed source", "type of germplasm storage code", "acquisition date", "location_code", "ploidy_level", "genome_structure", "ncbi_taxonomy_id", "transgenic", "introgression_parent", "introgression_backcross_parent", "introgression_map_version", "introgression_chromosome", "introgression_start_position_bp", "introgression_end_position_bp", "purdy_pedigree", "filial_generation")</code></td>
</tr>
<tr class="odd">
<td align="left">breedbase.cross_types</td>
<td align="left">The supported Pedigree cross types</td>
<td align="left"><code>c("biparental", "self", "open", "sib")</code></td>
</tr>
<tr class="even">
<td align="left">breedbase.country_codes</td>
<td align="left">List of 3-letter country codes</td>
<td align="left"><code>c('AFG','ALA','ALB','DZA','ASM','AND','AGO','AIA','ATA','ATG','ARG','ARM','ABW','AUS','AUT','AZE','BHS','BHR','BGD','BRB','BLR','BEL','BLZ','BEN','BMU','BTN','BOL','BES','BIH','BWA','BVT','BRA','IOT','VGB','BRN','BGR','BFA','BDI','CPV','KHM','CMR','CAN','CYM','CAF','TCD','CHL','CHN','HKG','MAC','CXR','CCK','COL','COM','COG','COK','CRI','CIV','HRV','CUB','CUW','CYP','CZE','PRK','COD','DNK','DJI','DMA','DOM','ECU','EGY','SLV','GNQ','ERI','EST','SWZ','ETH','FLK','FRO','FJI','FIN','FRA','GUF','PYF','ATF','GAB','GMB','GEO','DEU','GHA','GIB','GRC','GRL','GRD','GLP','GUM','GTM','GGY','GIN','GNB','GUY','HTI','HMD','VAT','HND','HUN','ISL','IND','IDN','IRN','IRQ','IRL','IMN','ISR','ITA','JAM','JPN','JEY','JOR','KAZ','KEN','KIR','KWT','KGZ','LAO','LVA','LBN','LSO','LBR','LBY','LIE','LTU','LUX','MDG','MWI','MYS','MDV','MLI','MLT','MHL','MTQ','MRT','MUS','MYT','MEX','FSM','MCO','MNG','MNE','MSR','MAR','MOZ','MMR','NAM','NRU','NPL','NLD','NCL','NZL','NIC','NER','NGA','NIU','NFK','MKD','MNP','NOR','OMN','PAK','PLW','PAN','PNG','PRY','PER','PHL','PCN','POL','PRT','PRI','QAT','KOR','MDA','REU','ROU','RUS','RWA','BLM','SHN','KNA','LCA','MAF','SPM','VCT','WSM','SMR','STP','SAU','SEN','SRB','SYC','SLE','SGP','SXM','SVK','SVN','SLB','SOM','ZAF','SGS','SSD','ESP','LKA','PSE','SDN','SUR','SJM','SWE','CHE','SYR','TJK','THA','TLS','TGO','TKL','TON','TTO','TUN','TUR','TKM','TCA','TUV','UGA','UKR','ARE','GBR','TZA','UMI','USA','VIR','URY','UZB','VUT','VEN','VNM','WLF','ESH','YEM','ZMB','ZWE')</code></td>
</tr>
<tr class="odd">
<td align="left">breedbase.location_types</td>
<td align="left">The supported types of Locations</td>
<td align="left"><code>c('Farm', 'Field', 'Greenhouse', 'Screenhouse', 'Lab', 'Storage', 'Other')</code></td>
</tr>
<tr class="even">
<td align="left">breedbase.design_types</td>
<td align="left">The supported Trial design types</td>
<td align="left"><code>c('CRD', 'RCBD', 'Alpha', 'Augmented', 'MAD', 'Westcott', 'Lattice')</code></td>
</tr>
<tr class="odd">
<td align="left">breedbase.trial_types</td>
<td align="left">The supported Trial types</td>
<td align="left"><code>c('Seedling Nursery', 'phenotyping_trial', 'Advanced Yield Trial', 'Preliminary Yield Trial', 'Uniform Yield Trial', 'Variety Release Trial', 'Clonal Evaluation', 'genetic_gain_trial', 'storage_trial', 'heterosis_trial', 'health_status_trial', 'grafting_trial', 'Screen House', 'Seed Multiplication', 'crossing_block_trial', 'Specialty Trial')</code></td>
</tr>
<tr class="even">
<td align="left">breedbase.accession_search_server</td>
<td align="left">The URL of the <a href="https://github.com/TriticeaeToolbox/BrAPI-germplasm-search" class="external-link">BrAPI germplasm search</a> server used by the Accession search functions</td>
<td align="left"><code>"https://synonyms.triticeaetoolbox.org"</code></td>
</tr>
<tr class="odd">
<td align="left">breedbase.accession_search_config</td>
<td align="left">The default Accession search parameters</td>
<td align="left"><code>list(database_terms = list(name = TRUE, synonyms = TRUE, accession_numbers = TRUE),search_routines = list(name = TRUE, punctuation = TRUE, substring = TRUE, edit_distance = FALSE, max_edit_distance = 2),return_records = FALSE)</code></td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="current-status">Current Status<a class="anchor" aria-label="anchor" href="#current-status"></a>
</h2>
<p>The package currently contains functions for the following data types:</p>
<ul>
<li>Accession</li>
<li>Pedigree</li>
<li>Location</li>
<li>Trial</li>
<li>Plot</li>
<li>PlotData</li>
<li>Conversions</li>
<li>Accession Searches (using the <a href="https://github.com/TriticeaeToolbox/BrAPI-germplasm-search" class="external-link">BrAPI-germplasm-search tool</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing breedbase</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>David Waring <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by David Waring.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
